[ntuple] Change entry handling in `RNTupleProcessor` to properly handle missing values

enirolf opened the pull request:
This PR adds a significant change to the way field values can be accessed from an `RNTupleProcessor`, and, under the hood, how the `RNTupleProcessor` handles the entries that manage these values. The motivation behind this change is that the current processor interface has two main shortcomings:
1. It is too flexible, allowing users to call `GetPtr` on the entry *inside* the event loop which is generally considered a bad practice.
2. It cannot properly handle values missing from entries as the result of an incomplete join or missing fields in subsequent chains.

To address these shortcomings, we introduce two new classes, the `RNTupleProcessorEntry` and the `RNTupleProcessorOptionalPtr`, which are described in more detail below. With the introduction of these classes, the use of the `RNTupleModel` is removed from the public interface entirely.

The changes in this PR are also in preparation of the use of the `RNTupleProcessor` as an RDataFrame data source. A minimal working example has been developed to validate that this new `RNTupleProcessor` interface is compatible with the abstract interface of `RDataSource`.

## The RNTupleProcessorEntry

The `RNTupleProcessorEntry` is an internal class that largely mirrors the `REntry` interface, but with additional functionality for the `RNTupleProcessor`. Most notably, fields and their values present in the entry have a notion of *validity*. Only when a field is valid, its value can be read.

In addition, it keeps track of which fields are auxiliary if a join operation is involved. This is needed in order to properly resolve from which underlying RNTuple the field data should be read. Since the RNTupleProcessor is composable, this is also relevant for the chain-based processor.

## The RNTupleProcessorOptionalPtr

The `RNTupleProcessorOptionalPtr` is what the user or upstream application will actually iteract with to read field values from a processor entry. It is obtained when registering a field in the processor through `RegisterField()`, which has to be done *before* and processor iteration starts (because, as stated before, afterwards the entry will be frozen).
The `RNTupleProcessorOptionalPtr` can be used to get a const reference or a pointer to the field's value. If the field is marked invalid in the entry at the time of access, the const reference access method will throw an exception, whereas the pointer access method returns a `nullptr`.

## Code example

### Old interface
```cpp
auto model = RNTupleModel::Create();
auto fldX = model->MakeField<float>("x");

auto proc =
  RNTupleProcessor::CreateChain({{"ntuple", "ntuple1.root"}, {"ntuple", "ntuple1.root"}}, std::move(model));

for (const auto &entry : *proc) {
  // Either (no way to guarantee the validity of `fldX` in this entry):
  std::cout << "x = " << *fldX << std::endl;
  // Or, same issue as above, plus expensive:
  std::cout << "x = " << entry.GetPtr<float>("x") << std::endl;
}
```

### New interface
```cpp
auto proc =
  RNTupleProcessor::CreateChain({{"ntuple", "ntuple1.root"}, {"ntuple", "ntuple1.root"}});

// Returns RNTupleProcessorOptionalPtr
auto x = proc->RequestField<float>("x");

// Instead of the full entry, now only provide the entry number as iterator value
for (const auto &idx : *proc) {
  // Without this check, `*x` will throw for entries where it is invalid.
  if (x.HasValue())
    std::cout << "x = " << *x << std::endl;
}
```

enirolf committed:
[ntuple] Use entry index as iterator value

...instead of the full entry. This design was already a bit
questionable, because it would allow users to call `GetPtr` or related
functions *inside* of the loop, which we want to avoid. Other than
that, there is (currently) no additional useful information one can
obtain from the full entry. Therefore, instead we now only return the
index of the current entry.

enirolf committed:
[ntuple] Update tutorials

enirolf committed:
[ntuple] Use specialized entry handling in `RNTupleProcessor`

After a chain or join operation, is is not guaranteed anymore that
fields always have a (valid) value in every entry. The existing `RENtry`
does not provide enough functionality to handle this, so instead we
implement an (internal) version of it (the `RNTupleProcessorEntry`),
which is specialized for the `RNTupleProcessor` to also keep track of
the validity of its values, as well as a class (the
`RNTupleProcessorOptionalPtr`) that can be used to read field values
from a processor with proper validity checks.

A major effect of this change is that the RNTupleModel is now hidden
from the user, i.e., it is not possible anymore to provide one through
the factory methods. The entry is instead filled on a field-by-field
basis, through the user-facing `RequestField` method. This method adds
the field to the entry (if not yet present), and returns an
`RNTupleProcessorOptionalPtr`. For advanced use cases, it is possible to
provide a raw pointer created by the user for reading the field values.
This introduces a risk of still reading wrong values when this pointer
is accessed directly, so the prescribed way to read entry data during
processing is always through the `RNTupleProcessorOptionalPtr`. This
feature will therefore not be publicly advertised and should only be
used by expert users.

enirolf committed:
[ntuple] Make number of entries in fixture multiple of 5

This will reduce the amount for entry number arithmetic in (future)
tests

pcanal commented on tree/ntuple/inc/ROOT/RNTupleProcessor.hxx:
```diff
@@ -189,14 +335,23 @@ public:
+   ///
+   /// \return An RNTupleProcessorOptionalPtr, which provides access to the field's value.
+   template <typename T>
+   RNTupleProcessorOptionalPtr<T> RegisterField(const std::string &fieldName)
```
Rather than 'registering' a (new) field, isn't this getting an accessor (or handle or proxy or ...)  to the field['s data].   Related ... how/when does it fail if the `fieldName` does not exist in the files?

enirolf replied:
It does both. By calling `RegisterField`, it will be added to the processor's entry so its values will actually be read and an `RNTupleProcessorOptionalPtr` to the field's entry value is returned. Open to alternative naming suggestions!

Thanks for raising the point about how to fail when the field doesn't actually exist, I didn't properly take this into account yet. It lead me to refactor the code in the PR more than expected, but I think this was a necessary step in the right direction. The behavior is now as follows, included in the tests:
https://github.com/root-project/root/blob/bc98934aaaf965750cb24c0174b3a4b19c4b5fca/tree/ntuple/test/ntuple_processor.cxx#L149-L156

pcanal replied:
Maybe rather than `Register` (and its association with creation), what about `RequestField` or simply `GetFieldValuePtr`?

hahnjo commented on tree/ntuple/inc/ROOT/RNTupleProcessor.hxx:
```diff
@@ -392,18 +528,16 @@ private:
    /// \brief Construct a new RNTupleProcessor for processing a single RNTuple.
    ///
    /// \param[in] ntuple The source specification (name and storage location) for the RNTuple to process.
-   /// \param[in] model The model that specifies which fields should be read by the processor.
    /// \param[in] processorName Name of the processor. Unless specified otherwise in RNTupleProcessor::Create, this is
    /// the name of the underlying RNTuple.
-   RNTupleSingleProcessor(RNTupleOpenSpec ntuple, std::unique_ptr<ROOT::RNTupleModel> model,
-                          std::string_view processorName);
+   RNTupleSingleProcessor(RNTupleOpenSpec ntuple, std::string_view processorName);
 
 public:
    RNTupleSingleProcessor(const RNTupleSingleProcessor &) = delete;
    RNTupleSingleProcessor(RNTupleSingleProcessor &&) = delete;
    RNTupleSingleProcessor &operator=(const RNTupleSingleProcessor &) = delete;
    RNTupleSingleProcessor &operator=(RNTupleSingleProcessor &&) = delete;
-   ~RNTupleSingleProcessor() override { fModel.release(); };
+   ~RNTupleSingleProcessor() override { fProtoModel.release(); };
```
Hm, why is `RNTupleSingleProcessor` releasing `RNTupleProcess::fProtoModel`? Does it have to do with the page source and destruction order? Is this still needed with the transition to a proto model?

pcanal replied:
'Why' is it needed?

enirolf replied:
Yeah I realized just now why: It's because otherwise the page source managed by the `RSingleProcessor` gets destructed first, and the fields in `fProtoModel` are connected to this page source, so it has to be destructed before.

hahnjo commented on tree/ntuple/inc/ROOT/RNTupleProcessorEntry.hxx:
```diff
@@ -0,0 +1,266 @@
+   /// The proto model contains all available fields. The entry itself will only contain a subset of these fields (i.e.,
+   /// the ones actually requested by the user).
+   ROOT::RNTupleModel *fProtoModel;
+   /// Corresponds to the fields of the proto model
+   std::vector<ROOT::RFieldBase::RValue> fValues;
+   /// Marks whether fields are valid for reading
+   std::vector<bool> fFieldValidities;
+   /// For fast lookup of token IDs given a (sub)field name present in the entry
+   std::unordered_map<std::string, std::size_t> fFieldName2Token;
```
Missing a couple of includes:
```c++
#include <ROOT/RFieldBase.hxx>
#include <ROOT/RFieldToken.hxx>

#include <cassert> // for assert() below
#include <cstddef> // for std::size_t
#include <string>
#include <string_view>
#include <unordered_map>
#include <vector>
```

hahnjo commented on tree/ntuple/inc/ROOT/RNTupleProcessorEntry.hxx:
```diff
@@ -0,0 +1,266 @@
+   /// `false` otherwise.
+   ///
+   /// A field will only be updated if `valuePtr` is not a `nullptr`.
+   bool AddOrUpdateField(const std::string &fieldName, std::shared_ptr<void> valuePtr)
```
```suggestion
   bool AddOrUpdateField(std::string_view fieldName, std::shared_ptr<void> valuePtr)
```
(unless there is a good reason to force a `std::string`; `FindToken` itself accepts a `std::string_view` and will convert it to `std::string` for searching in `fFieldName2Token`...)

pcanal replied:
Hopefully that is not a performance critical path, otherwise it seems 'wasteful' to pass an `std::string` via `std::string_view` to turn around and make a duplicate of the same `std::string` (unless there is some transformation in between).

hahnjo commented on tree/ntuple/inc/ROOT/RNTupleProcessorEntry.hxx:
```diff
@@ -0,0 +1,266 @@
+      assert(token.fSchemaId == fProtoModel->fSchemaId && token.fIndex < fValues.size());
+
+      if (fFieldValidities[token.fIndex])
+         return std::static_pointer_cast<T>(fValues[token.fIndex].GetPtr<void>());
```
Isn't this potentially dangerous without a type check? Should it do
```suggestion
         return fValues[token.fIndex].GetPtr<T>();
```
(knowing that right now, `RValue::GetPtr` doesn't do a type check either - we should fix https://github.com/root-project/root/issues/18316...)

hahnjo commented on tree/ntuple/src/RNTupleProcessor.cxx:
```diff
@@ -477,61 +400,39 @@ void ROOT::Experimental::RNTupleJoinProcessor::SetModel(std::unique_ptr<ROOT::RN
-                                  fAuxiliaryProcessor->GetProcessorName() + "\""));
-      }
+      if (fAuxiliaryProcessor->LoadEntry(entryNumber) == kInvalidNTupleIndex)
+         SetAuxiliaryFieldValidity(false);
```
Where is the inverse happening, ie where are we setting fields as valid if the next entry exists again?

hahnjo commented on tree/ntuple/src/RNTupleProcessor.cxx:
```diff
@@ -259,36 +205,25 @@ void ROOT::Experimental::RNTupleSingleProcessor::PrintStructureImpl(std::ostream
-   }
+void ROOT::Experimental::RNTupleChainProcessor::Initialize()
+{
+   // The processor has already been initialized.
+   if (fProtoModel && fEntry)
+      return;
```
```suggestion
   if (IsInitialized())
      return;
```

hahnjo reviewed (commented):
First code review - I have two overall design questions that I will post separately so their inline comments are not mixed with these.

hahnjo reviewed (commented):
My two questions:
1. Do we (want to) support conversion and schema evolution when reading fields via the `RNTupleProcessor`? (triggered by https://github.com/root-project/root/pull/19111#discussion_r2194261277)
2. At the moment, every `RNTupleProcessor` has its own `RNTupleProcessorEntry` that are "connected together" via shared pointers. However, the field validity is not shared which could lead to problems with nested joins where the outer one is valid, but the inner one does not find its value (see also https://github.com/root-project/root/pull/19111/files#r2194283077).

pcanal commented:
> My two questions:
> 
>     1. Do we (want to) support conversion and schema evolution when reading fields via the `RNTupleProcessor`? (triggered by [#19111 (comment)](https://github.com/root-project/root/pull/19111#discussion_r2194261277))
 
In my opinion we have no choice but to (eventually) support schema evolution (otherwise the  `RNTupleProcessor` would be usable to read any non-current files).

enirolf reviewed (commented):
Thank you for you extensive review, @hahnjo! To answer your questions:

> 1. Do we (want to) support conversion and schema evolution when reading fields via the RNTupleProcessor? (triggered by https://github.com/root-project/root/pull/19111#discussion_r2194261277)

Yes, to echo what Philippe wrote, we have no choice (not just for schema evolution, but also to allow alternative field types through e.g., RDF). I admit that I didn't yet take this into account in this particular PR, but I also think what with the proposed interface change there's no blocker not to have this. I will address it and depending on how much it would blow up this PR even more, I'll either add it here or in a follow-up PR.

> 2. At the moment, every RNTupleProcessor has its own RNTupleProcessorEntry that are "connected together" via shared pointers. However, the field validity is not shared which could lead to problems with nested joins where the outer one is valid, but the inner one does not find its value (see also https://github.com/root-project/root/pull/19111/files#r2194283077).

That's a good point, and something that needs to indeed also be addressed more properly. For this one as well, I will investigate more and either add it to this PR or create a follow-up.

vepadulano commented on tree/ntuple/test/ntuple_processor.cxx:
```diff
@@ -143,79 +142,44 @@ TEST_F(RNTupleProcessorTest, Base)
-      std::vector<float> yExp{static_cast<float>(nEntries - 1), static_cast<float>((nEntries - 1) * 2)};
-      EXPECT_EQ(yExp, *entry.GetPtr<std::vector<float>>("y"));
+   try {
+      proc->RegisterField<float>("z");
+      FAIL() << "registering fields that do not exist should not be possible";
```
This may not need be addressed in this PR, but in general this is possible. See e.g. https://github.com/root-project/root/blob/00565c678832d07c6c73cd2d732236d11c87b710/tree/dataframe/test/dataframe_incomplete_entries.cxx#L220

Users processing a vertical concatenation of N files may not know if field "x" is always present in all files, and in an unlucky scenario the field may be missing exactly from the first file.

enirolf replied:
I see. In principle there's nothing holding us back from allowing users to request arbitrary fields and have their validity set to `false` until it is actually found on disk. I will look into this possibility further and address it in a separate PR.

vepadulano reviewed (changes requested):
The work is already at a really advanced state and the direction is very positive, thanks! A few minor suggestions/considerations. ~~Furthermore, I don't see the new API to check for missing values e.g. FieldIsValid` being used in the tests, would be good to have at least one simple test already in this PR. A more thorough testing campaign could be added later.~~

EDIT: Sorry, I was looking only for `FieldIsValid`, I see there are uses of `HasValue` in the tests :+1:

jblomer commented on tree/ntuple/inc/ROOT/RNTupleProcessor.hxx:
```diff
@@ -189,14 +342,31 @@ public:
-   const ROOT::REntry &GetEntry() const { return *fEntry; }
+   /// \return An RNTupleProcessorOptionalPtr, which provides access to the field's value.
+   template <typename T>
+   RNTupleProcessorOptionalPtr<T> RegisterField(std::string_view fieldName)
```
I think I'd rather mimic the view API here, i.e. `GetView()` returning an object that probably derives from `RNTupleViewBase`.

jblomer reviewed (commented):
I think the tutorials look much cleaner with this change!

pcanal commented on tutorials/io/ntuple/ntpl015_processor_join.C:
```diff
@@ -77,23 +77,25 @@ void Read()
+   // Access to the processor's fields is done by first registering them through RNTupleProcessor::RegisterField(). The
+   // returned value can be used to read the current entry's value for that particular field. Fields from the primary
+   // ntuple are registered by their original name.
+   auto px = processor->RegisterField<float>("vpx");
```
```suggestion
   auto px = processor->RequestField<float>("vpx");
```
which may (or may not) closer to the actual semantic of the function.

enirolf committed:
[ntuple] Add `RNTupleProcessorEntry`

Interal class that largely mirrors the REntry interface, but with
additional functionality specifically for the `RNTupleProcessor`. Most
notably, fields and their values present in the entry have a notion of
validity. Only when a field is valid, its value can be read.

enirolf committed:
[ntuple] Add `RNTupleProcessorProvenance`

This tracks the provenance of a field through composed processors, so
its data is correctly read from the actual on-disk field.

enirolf committed:
[ntuple] Add more tests for composed processors

hahnjo reviewed (commented):
I looked through the code and this all looks very reasonable to me, only a couple minor comments inline.

I agree that splitting "Use specialized entry handling in `RNTupleProcessor`" is hard. I have two ideas, but to be clear I'm not sure it's worth the effort to actually put them in:
1. It feels like half the modifications in the tests are changing `GetEntry().GetPtr<T>("...")` to `RequestField<T>("...")`. To reduce the diff, you could introduce `RequestField` as an alias first and apply the test changes beforehand.
2. Quite some changes in the `RNTupleProcessor.hxx` are for the removal of the `model` arguments. It could be possible to just leave them in for one commit (but ignore) and remove them afterwards.

So in effect it should be possible to split into three commits:
1. `GetEntry().GetPtr<T>("...")` -> `RequestField<T>("...")`
2. The "rest" of "Use specialized entry handling in `RNTupleProcessor`"
3. Remove `model` arguments from constructors and factory functions

Again, not sure if it's actually worth the effort to put it in now.

enirolf committed:
[ntuple] Change processor iterator tag

See commit e5fba74 for the "why".

enirolf reviewed (commented):
Thanks @hahnjo! If the other reviewers prefer it I can apply your suggestion to split the commit, otherwise I propose to keep it as-is.

silverweed commented on tree/ntuple/inc/ROOT/RNTupleProcessor.hxx:
```diff
@@ -125,15 +277,41 @@ protected:
+   /// \brief Check if a field exists on-disk and can be read by the processor.
+   ///
+   /// \param[in] fieldName Name of the field to check.
+   virtual bool FieldExists(std::string_view fieldName) = 0;
```
Another idea would be `CanRead(Field)FromDisk`, but it starts getting long. I would probably mention `Disk` somewhere in the name though

vepadulano commented on tree/ntuple/inc/ROOT/RNTupleProcessor.hxx:
```diff
@@ -65,13 +65,142 @@ public:
+   }
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Get a raw pointer to the field value managed by the processor's entry.
```
```suggestion
   /// \brief Get a non-owning pointer to the field value managed by the processor's entry.
```

vepadulano commented on tree/ntuple/inc/ROOT/RNTupleProcessor.hxx:
```diff
@@ -65,13 +65,142 @@ public:
+   /// \brief Bind the value to `valuePtr`.
+   ///
+   /// \param[in] valuePtr Pointer to bind the value to.
+   void BindRawPtr(T *valuePtr) { fProcessorEntry->BindRawPtr(fFieldIndex, valuePtr); }
```
I'm trying to think about what use case this serves. `RNTupleProcessorOptionalPtr` is a public class and this is a public method. So I am inferring that the RNTupleProcessorOptionalPtr can be used as a detached object that binds its internal entry to a raw pointer? But then, the internal entry is internal, so only an RNTupleProcessor can set it, so calling `BindRawPtr` from the user side may result in accessing invalid memory. This is just free-flow of thoughts, maybe there is some chance of clarifying the intent here.

enirolf replied:
You are right, but to also allow the processor to be used in core frameworks, during an earlier offline discussion we came to the conclusion that it has to be possible for it to read into user-provided pointers. It is true that this comes with the risk of reading invalid values and the responsibility to prevent this is moved to the user (which can still be achieved by only reading through the `RNTupleProcessorOptionalPtr`). This is an interface for advanced use only and will not be advertised. Thinking about it again we could consider moving this functionality to a free function in the Detail namespace, but then we also need to change again the option to pass a pointer to `RequestField`. I will add a clarifying comment for now.

vepadulano reviewed (changes requested):
Thank you! This is already very mature and close to being ready. A few comments from my side before the final push.

enirolf commented on tree/ntuple/src/RNTupleProcessor.cxx:
```diff
@@ -129,108 +101,138 @@ std::unique_ptr<ROOT::Experimental::RNTupleProcessor> ROOT::Experimental::RNTupl
+   for (const auto &fieldIdx : fFieldIdxs) {
+      const auto &entryField = fEntry->GetField(fieldIdx);
+
+      // TODO handle subfields
```
Yes, but I already have a follow-up PR lined up to address this, it goes hand-in-hand with the possibility to add fields with alternative types that's too big IMO to still include in this one.

enirolf commented on tree/ntuple/test/ntuple_processor.cxx:
```diff
@@ -143,79 +157,69 @@ TEST_F(RNTupleProcessorTest, Base)
-      EXPECT_EQ(nEntries - 1, proc->GetCurrentEntryNumber());
+   try {
+      proc->RequestField<float>("z");
+      FAIL() << "registering fields that do not exist should not be possible";
```
I am aware, and will address this in a follow-up!

vepadulano commented on tree/ntuple/src/RNTupleProcessor.cxx:
```diff
@@ -129,108 +101,138 @@ std::unique_ptr<ROOT::Experimental::RNTupleProcessor> ROOT::Experimental::RNTupl
+         fieldIdx = fEntry->AddField(fieldName, field, valuePtr, provenance);
+         return *fieldIdx;
+      } catch (const ROOT::RException &) {
+         return R__FAIL("cannot register field with name \"" + std::string(fieldName) +
```
Neat!

vepadulano reviewed (approved):
Great work, thank you! Please try to simplify the commit history where it makes sense before merging.

hahnjo reviewed (approved):
My last round of comments has been addressed :smiley:

enirolf commented on tree/ntuple/inc/ROOT/RNTupleProcessor.hxx:
```diff
@@ -330,17 +524,12 @@ public:
-              std::unique_ptr<ROOT::RNTupleModel> auxModel = nullptr, std::string_view processorName = "");
+   static std::unique_ptr<RNTupleProcessor> CreateJoin(RNTupleOpenSpec primaryNTuple, RNTupleOpenSpec auxNTuple,
+                                                       const std::vector<std::string> &joinFields,
+                                                       std::string_view processorName = "");
```
This can be achieved by composing different joins, see e.g. https://github.com/enirolf/root/blob/9e7877f309315a0327322741a965d3e9670d72c6/tree/ntuple/test/ntuple_processor.cxx#L434-L458.

In the original design the `RNTupleJoinProcessor` could be constructed with multiple auxiliary RNTuples/inner processors, but this made the interface kind of clunky and inconvenient for the what I currently believe will be the most common use case, which is a join between just 2 datasets.

jblomer commented on tree/ntuple/inc/ROOT/RNTupleProcessorEntry.hxx:
```diff
@@ -0,0 +1,303 @@
+      }
+      // Should never happen, but avoid compiler warning about "returning reference to local temporary object".
+      static const std::string empty = "";
+      return empty;
```
Possibly: `R__ASSERT(false)`

jblomer reviewed (approved):
Very nice!
